src/interfaces.ts

```diff
@@ -112,20 +112,35 @@ export interface CollectionRef {
 
 export type PropertyFilterOperator = '<' | '<=' | '>' | '>=' | ':' | '!:' | '=' | '!=';
 
+export interface PropertyFilterOperatorExtended<TokenValue> {
+  operator: PropertyFilterOperator;
+  match?: PropertyFilterOperatorMatch<TokenValue>;
+}
+
+export type PropertyFilterOperatorMatch<TokenValue> =
+  | PropertyFilterOperatorMatchByType
+  | PropertyFilterOperatorMatchCustom<TokenValue>;
+
+export type PropertyFilterOperatorMatchByType = 'date' | 'datetime';
+
+export type PropertyFilterOperatorMatchCustom<TokenValue> = (itemValue: unknown, tokenValue: TokenValue) => boolean;
+
 export type PropertyFilterOperation = 'and' | 'or';
 export interface PropertyFilterToken {
-  value: string;
+  // By default, the token value is a string.
+  // When a custom property is used, the token value can be any;
+  value: any;
```

Comment:
The token value is any due to challenges in passing is as generic.

Reply:
I'd like to understand the challenges to make the PropertyFilterToken value generic?  Could you elaborate?

Reply:
The tokens are used in the PropertyFilterQuery:

```
interface PropertyFilterQuery {
  operation: "and" | "or";
  tokens: readonly PropertyFilterToken[];
}
```

Because we have an array of tokens we can't really parametrise it as the tokens might have different type depending on the property key. It is theoretically possible to achieve if we'd have a mapping between property key and token value provided explicitly.